### PR TITLE
Mark State.setState as protected

### PR DESCRIPTION
**Pull request by abarth**
This required refactoring some cases where we weren't following the rules for
the protected annotation.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie @HansMuller

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 1]
```diff
@@ -612,13 +612,13 @@ class GalleryDrawer extends StatelessWidget {
           new DrawerItem(
             icon: Icons.hourglass_empty,
             selected: timeDilation != 1.0,
-            onPressed: () { _toggleAnimationSpeed(context); },
+            onPressed: () { ComplexLayoutApp.of(context).toggleAnimationSpeed(); },
```
Comment (Hixie) [inline dev/benchmarks/complex_layout/lib/main.dart]: we normally use callbacks for this, but ok

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 2]
```diff
@@ -64,13 +64,13 @@ class OverlayEntry {
     if (_opaque == value)
       return;
     assert(_overlay != null);
-    _overlay.setState(() {
+    _overlay._setEntryOpacity(() {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/overlay.dart]: With that name, I think I'd pass a value and have _setEntryOpacity set _opaque on the overlay entry itself.

File: packages/flutter/lib/src/widgets/routes.dart [hunk 3]
```diff
@@ -515,7 +517,7 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
     if (_offstage == value)
       return;
     _offstage = value;
-    _scopeKey.currentState?.setState(() {
+    _scopeKey.currentState?._setRouteOffstage(() {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/routes.dart]: call this _didSetRouteOfffStage, no arguments

File: packages/flutter/test/widget/build_scope_test.dart [hunk 4]
```diff
@@ -52,6 +52,10 @@ class BadWidgetParent extends StatefulWidget {
 }
 
 class BadWidgetParentState extends State<BadWidgetParent> {
+  void _markNeedsBuild() {
+    setState(() {});
```
Comment (Hixie) [inline packages/flutter/test/widget/build_scope_test.dart]: comment in the closure

**Comment by Hixie** [comment]
So... should we make markNeedsBuild @protected? :-)

LGTM.

**Pull request merged**